extensions/amp-video-docking/1.0/dock.js

```diff
@@ -0,0 +1,488 @@
+/**
+ * Copyright 2020 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as Preact from '../../../src/preact';
+import {
+  CORNER_MARGIN_MAX,
+  CORNER_MARGIN_RATIO,
+  CORNER_WIDTH_MIN,
+  CORNER_WIDTH_RATIO,
+  DirectionX,
+  DirectionY,
+  DockTargetType,
+  PLACEHOLDER_ICON_LARGE_MARGIN,
+  PLACEHOLDER_ICON_LARGE_WIDTH,
+  PLACEHOLDER_ICON_SMALL_MARGIN,
+  PLACEHOLDER_ICON_SMALL_WIDTH,
+  REVERT_TO_INLINE_RATIO,
+  transform,
+} from '../0.1/def';
+import {
+  calculateLeftJustifiedX,
+  calculateRightJustifiedX,
+  interpolatedBoxesTransform,
+  isSizedRect,
+  topCornerRect,
+} from '../0.1/math';
+import {createContext, useMemo, useState} from '../../../src/preact';
+import {dict} from '../../../src/utils/object';
+import {once} from '../../../src/utils/function';
+import {useStyles} from './dock.jss';
+import {Controls} from './controls';
+import {Placeholder} from './placeholder';
+
+/**
+ * @param {!Element} element
+ * @return {!Promise<number>}
+ */
+const getOffsetTopAsync = (element) =>
+  Promise.resolve().then(() => element.offsetTop);
+
+// TODO: i don't like this name
+class DockManager {
+  /**
+   * @param {!Document} doc
+   * @param {function(!JsonObject)} setState
+   * @param {function(!JsonObject)} setStyles
+   */
+  constructor(doc, setState, setStyles) {
+    this.doc_ = doc;
+
+    /** @private {?{entry: !IntersectionObserverEntry, rect: !LayoutRectDef}} */
+    this.current_ = null;
+
+    /** @private @const {function(!JsonObject)} */
+    this.setState_ = setState;
+
+    /** @private @const {function(!JsonObject)} */
+    this.setStyles_ = setStyles;
```

Comment:
`state` and `styles` should probably be merged into the same object—haven't gotten there yet since I'd like to confirm re-renders.